### PR TITLE
MINOR: Guard against NPE when throwing StreamsException on serializer mismatch

### PR DESCRIPTION
**Pull request by miguno**

### CODE REVIEW

**Comment by miguno** [comment]
Please review any of @dguy @enothereska @guozhangwang @mjsax.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java [hunk 1]
```diff
@@ -77,14 +77,16 @@ public void process(final K key, final V value) {
         try {
             collector.send(topic, key, value, null, timestamp, keySerializer, valSerializer, partitioner);
         } catch (ClassCastException e) {
+            String keyClass = key == null? "unknown because key was null": key.getClass().getName();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java]: nit: `final` 
also the next line

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java [hunk 2]
```diff
@@ -69,4 +68,52 @@ public void shouldThrowStreamsExceptionOnKeyValyeTypeSerializerMissmatch() {
         }
     }
 
+    @Test(expected = StreamsException.class)
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java]: remove this and just put a `fail(...)` inside the `try{..}catch`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java [hunk 3]
```diff
@@ -69,4 +68,52 @@ public void shouldThrowStreamsExceptionOnKeyValyeTypeSerializerMissmatch() {
...
+            new RecordCollectorImpl(new MockProducer<byte[], byte[]>(true, anySerializer, anySerializer),null));
+        context.setTime(1);
+        final SinkNode sink = new SinkNode<>("anyNodeName", "any-output-topic", anySerializer, anySerializer, null);
+        sink.init(context);
+
+        String invalidKeyToTriggerSerializerMismatch = "";
+
+        // When/Then
+        try {
+            sink.process(invalidKeyToTriggerSerializerMismatch, null);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java]: i.e., add `fail` after this line

**Review by dguy — COMMENTED** [review]
Thanks @miguno i've left some comments

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java [hunk 4]
```diff
@@ -47,12 +48,10 @@ public void invalidInputRecordTimestampTest() {
 
     @Test(expected = StreamsException.class)
     @SuppressWarnings("unchecked")
-    public void shouldThrowStreamsExceptionOnKeyValyeTypeSerializerMissmatch() {
+    public void shouldThrowStreamsExceptionOnKeyValueTypeSerializerMismatch() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java]: Can you update this, and the above test, too -> remove `expected = ...` and use `fail`

**Review by mjsax — COMMENTED** [review]
What @dguy said. :) Also as you already touching touching the test, please update all methods to use the same pattern.

**Comment by miguno** [comment]
Updated the PR and addressed review comments.  Still have to verify checkstyle is happy.

**Comment by miguno** [comment]
Checkstyle is happy now, too.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java [hunk 5]
```diff
@@ -20,52 +20,104 @@
...
 
-import java.util.Properties;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.fail;
 
 public class SinkNodeTest {
 
     @Test(expected = StreamsException.class)
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/SinkNodeTest.java]: Would you mind "fixing" this test to? i.e., using `@Test(expected = SomeException.class)` is bad form for tests with more than a single line. The exception could potentially happen anywhere and my be masking actual bugs. In these cases it is always better to use the same pattern as below, i.e., 
```
try {
    performExceptionThrowingAction()
    fail("KABOOM! shouldn't get here");
} catch (SomeException e) {
   // pass
}
```

**Review by dguy — COMMENTED** [review]
Just left a comment about updating one other test. Thanks @miguno

**Comment by miguno** [comment]
@dguy updated

**Review by dguy — APPROVED** [review]
Thanks @miguno LGTM

**Review by guozhangwang — APPROVED** [review]
Merged to trunk.

**Pull request closed without merging**